Documentation overhaul (#266)
nix-community/nix-doom-emacs: docs/README.md

# Documentation for Nix-Doom-Emacs

> Nix-Doom-Emacs (also commonly referred to as NDE in chatrooms) is a project with a lot of moving pieces and hacks; it's a project that is itself a fork of another project that has these moving pieces. Thus, it's not out of the blue to expect it to be a tiny bit more buggy for most people than other software. If you encounter any issues that make it not usable to you (or if you need support), please talk to us first in the [Matrix room](https://matrix.to/#/#doom-emacs:nixos.org) and if it's indeed a bug of Nix-Doom-Emacs, file it in the [issue tracker](https://github.com/nix-community/nix-doom-emacs/issues). If you find this documentation unclear or incomplete, please let us know as well.

Reviewer: ```suggestion
Nix-Doom-Emacs (also commonly referred to as NDE in chatrooms) is a project with lots of moving pieces and hacks. It is also not for the faint of heart, in general assuming that the user has experience with **both** Nix and Elisp (Emacs Lisp) code, and has a fairly amount of experience debugging issues.

If you encounter any issues that make it not usable to you (or if you need support), please talk to us first in the [Matrix room](https://matrix.to/#/#doom-emacs:nixos.org) and if it's indeed a bug of Nix-Doom-Emacs, file it in the [issue tracker](https://github.com/nix-community/nix-doom-emacs/issues). If you find this documentation unclear or incomplete, please let us know as well.
```

Here's the [FAQ](./faq.md)

> Nix-Doom-Emacs uses [nix-straight.el](https://github.com/nix-community/nix-straight.el) under the hood to install dependencies. It's a low level wrapper to add Nix integration over straight.el, the declarative package manager used by Doom Emacs. 

Reviewer: ```suggestion
Nix-Doom-Emacs uses [nix-straight.el](https://github.com/nix-community/nix-straight.el) under the hood to install dependencies. It's a low level wrapper to add Nix integration over [straight.el](https://github.com/radian-software/straight.el), the declarative package manager used by Doom Emacs. Before starting, please read its [README](https://github.com/nix-community/nix-straight.el/blob/master/README.md) file and understand its consequences.
```

If you're not aware yet, then:

> #### **WARNING**: HERE BE DRAGONS! THIS IS A FRAGILE PROJECT

Reviewer: ```suggestion
#### **WARNING**: HERE BE DRAGONS! THIS IS A PROJECT RECOMMEND ONLY FOR ADVANCED USERS
```

  };

  outputs = {
    self,
    nixpkgs,
    lib,
    home-manager,
    nix-doom-emacs,
    ...
  }:
    let
      system      = "x86_64-linux";
      specialArgs = { inherit inputs; };
    in {
    nixosConfigurations.exampleHost = lib.nixosSystem {
      inherit system specialArgs;
      modules = [
        ./default.nix
        home-manager.nixosModules.home-manager
        {
          home-manager.users.exampleUser.imports = [ ./home.nix ];
        }
      ];
    };
  };
}

```

`File: home.nix`
```nix
{ config, pkgs, inputs, ... }: {
  imports = [ inputs.nix-doom-emacs.hmModule ];

  # ...
  programs.doom-emacs = {
    enable = true;
    doomPrivateDir = ./doom.d; # Directory containing your config.el, init.el
                               # and packages.el files
  };
  # ...
}
```

## Non-Flake Home-Manager

```nix
{ pkgs, ... }:

let
  doom-emacs = pkgs.callPackage (builtins.fetchTarball {
    url = https://github.com/nix-community/nix-doom-emacs/archive/master.tar.gz;
  }) {
    doomPrivateDir = ./doom.d;  # Directory containing your config.el, init.el
                                # and packages.el files
  };
in {
  home.packages = [ doom-emacs ];
}
```

> ## Standalone

Reviewer: ```suggestion
## NixOS
```

I think we also need a "standalone" section, but this would give as an example a `shell.nix` file.

Reviewer 2: I'm not quite sure how a `flake.nix` file with a `devShell` for standalone systems would look like, would you provide an example snippet?

Reviewer: For `shell.nix`:

```
{ pkgs ? import <nixpkgs> { } }:

let
  repo = pkgs.fetchFromGitHub {
    owner = "nix-community";
    repo = "nix-doom-emacs";
    rev = "<commit>";
    sha256 = "<hash>";
  };
  nix-doom-emacs = pkgs.callPackage (import repo) {
    doomPrivateDir = ./doom.d;
  };
in
pkgs.mkShell {
  buildInputs = [ nix-doom-emacs ];
}
```

Reviewer: For `flake.nix`:

```nix
{
  description = "nix-doom-emacs shell";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
    nix-doom-emacs.url = "github:nix-community/nix-doom-emacs";
  };

  outputs = { self, nixpkgs, nix-doom-emacs, ... }:
  let
    system = "x86_64-linux";
    pkgs = import nixpkgs { inherit system; };
    doom-emacs = nix-doom-emacs.packages.${system}.default.override {
      doomPrivateDir = ./doom.d;
    };
  in
  {
    devShells.${system}.default = pkgs.mkShell {
      buildInputs = [ doom-emacs ];
    };
  };
}
```

Using Nix-Doom-Emacs standalone isn't recommended, especially if you're a beginner.

`File: flake.nix`
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    nix-doom-emacs.url = "github:nix-community/nix-doom-emacs";
  };
  
  outputs = {
    self,
    nixpkgs,
    lib,
    nix-doom-emacs,
    ...
  }:
    let
      system      = "x86_64-linux";
      specialArgs = { inherit inputs; };
    in {
    nixosConfigurations.exampleHost = lib.nixosSystem {
      inherit system specialArgs;
      modules = [
        ./default.nix
        # ...
      ];
    };
  };
}
```

`File: default.nix`
```nix
{ config, nixpkgs, lib, inputs }: {
  # ...
  environment.systemPackages = 
    let

>       doom-emacs = inputs.nix-doom-emacs.packages.${system}.default.override {

Reviewer: Can you add a comment here explaining that you can see all supported parameters inside our `default.nix` file?

# Documentation for nix-doom-emacs

nix-doom-emacs (also commonly referred to as NDE in chatrooms) is a project with lots of moving pieces and hacks. Users are expected to know their way around using (and especially debugging) Nix and Emacs Lisp before using this project.

If you encounter any issues that make it unusable to you (or if you need support), please talk to us first in the [Matrix room](https://matrix.to/#/#doom-emacs:nixos.org) and if it's indeed a bug of nix-doom-emacs, file it in the [issue tracker](https://github.com/nix-community/nix-doom-emacs/issues).

If you find this documentation unclear or incomplete, please let us know as well.

Here's the [FAQ](./faq.md)

> nix-doom-emacs uses [`nix-straight.el`](https://github.com/nix-community/nix-straight.el) under the hood to install dependencies. It's a low level wrapper to add Nix integration over [`straight.el`](https://github.com/radian-software/straight.el), the declarative package manager used by Doom Emacs. 

Reviewer 2: The "declarative pcakage manager used by Doom Emacs" isn't referring to Nix, it's referring to straight.el, but yeah, you're right

Before using nix-doom-emacs, make sure to read [`nix-straight.el`'s README'](https://github.com/nix-community/nix-straight.el), and that you understand the consequences.

# Getting Started

> To get started, we suggest these methods. They are ordered from most suggested to least suggested.

Reviewer 2: do you think it's implied? i'd really like to keep it in to at least give users some semblance of knowing what's most recommended

Before using Nix-Doom-Emacs, make sure to read [`nix-straight.el`'s README'](https://github.com/nix-community/nix-straight.el), and that you understand the consequences.

# Getting Started

To get started, we suggest these methods. They are ordered from most suggested to least suggested.
In all of these methods, you'll need your Doom Emacs configuration. It should contain the following three files: 
`config.el`, `init.el` and `packages.el`. If you don't already have an existing `doom-emacs` configuration, you can use the contents of `test/doom.d` as a template.

The Doom configuration will be referred to as `./doom.d` in these snippets. You can name it whatever you like.

## Flake + Home-Manager

`File: flake.nix`
```nix
{
  inputs = {
    home-manager.url = "github:nix-community/home-manager";
    nix-doom-emacs.url = "github:nix-community/nix-doom-emacs";
  };

  outputs = {
    self,
    nixpkgs,
    lib,
    home-manager,
    nix-doom-emacs,
    ...
  }:
    let
      system      = "x86_64-linux";

>       specialArgs = { inherit inputs; };

Reviewer: ```suggestion
```

This is not being used here, and I think it doesn't make sense to confuse folks about `specialArgs`, getting at the risk that maybe will think this is essential for it to work.

Reviewer 2: It's put there because it allows `home.nix` to get access to the `inputs` from the flake, which is something I think is very commonly needed, but I think I'll remove it since this isn't about getting the user an opinionated config

Reviewer: Yeah, I know that passing `inputs` as `specialArgs` is common. I just don't want folks assuming that this is essential to get it working.

    in {
    nixosConfigurations.exampleHost = lib.nixosSystem {

>       inherit system specialArgs;

Reviewer: ```suggestion
      inherit system;
```

      modules = [
        ./default.nix
        home-manager.nixosModules.home-manager
        {
          home-manager.users.exampleUser.imports = [ ./home.nix ];
        }
      ];
    };
  };
}

```

`File: home.nix`
```nix
{ config, pkgs, inputs, ... }: {
  imports = [ inputs.nix-doom-emacs.hmModule ];

  # ...
  programs.doom-emacs = {
    enable = true;
    doomPrivateDir = ./doom.d; # Directory containing your config.el, init.el
                               # and packages.el files
  };
  # ...
}
```

## Non-Flake Home-Manager

```nix
{ pkgs, ... }:

let
  doom-emacs = pkgs.callPackage (builtins.fetchTarball {
    url = https://github.com/nix-community/nix-doom-emacs/archive/master.tar.gz;
  }) {
    doomPrivateDir = ./doom.d;  # Directory containing your config.el, init.el
                                # and packages.el files
  };
in {
  home.packages = [ doom-emacs ];
}
```

## NixOS

> Using Nix-Doom-Emacs without Home-Manager isn't recommended, especially if you're a beginner.

Reviewer: Home-Manager really doesn't bring anything essential for NDE usage, so.
```suggestion
```

Reviewer: And we already made clear at the top that this list is ordered in the most preferred way first.

`File: flake.nix`
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    nix-doom-emacs.url = "github:nix-community/nix-doom-emacs";
  };
  
  outputs = {
    self,
    nixpkgs,
    lib,
    nix-doom-emacs,
    ...
  }:
    let
      system      = "x86_64-linux";
      specialArgs = { inherit inputs; };
    in {
    nixosConfigurations.exampleHost = lib.nixosSystem {
      inherit system specialArgs;
      modules = [
        ./default.nix
        # ...
      ];
    };
  };
}
```

`File: default.nix`
```nix
{ config, nixpkgs, lib, inputs }: {
  # ...
  environment.systemPackages = 
    let
      doom-emacs = inputs.nix-doom-emacs.packages.${system}.default.override {
        doomPrivateDir = ./doom;
      };
    in [
      doom-emacs
    ];
  # ...
}
```

For what it's worth, you can see all overridable parameters of Nix-Doom-Emacs in [default.nix](../default.nix).

## Standalone

> This is the least recommended method. This uses the `devShell` (or `nix-shell` feature if you're using non-flakes, which is not recommended) feature of Nix.

Reviewer: Same as the NixOS comment.
```suggestion
```
